src/server/frontend_wayland/idle_inhibit_v1.cpp

```diff
@@ -0,0 +1,173 @@
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: Grayson Guarino <grayson.guarino@canonical.com>
+ */
+
+#include "idle_inhibit_v1.h"
+
+#include "mir/log.h"
+#include "mir/executor.h"
+#include "mir/scene/idle_hub.h"
+
+#include <boost/throw_exception.hpp>
+#include <memory>
+
+namespace mf = mir::frontend;
+namespace mw = mir::wayland;
+namespace ms = mir::scene;
+
+namespace mir
+{
+namespace frontend
+{
+struct IdleInhibitV1Ctx
+{
+    std::shared_ptr<Executor> const wayland_executor;
+    std::shared_ptr<scene::IdleHub> const idle_hub;
+};
+
+class IdleInhibitManagerV1Global
+    : public wayland::IdleInhibitManagerV1::Global
+{
+public:
+    IdleInhibitManagerV1Global(wl_display* display, std::shared_ptr<IdleInhibitV1Ctx> const& ctx);
+
+private:
+    void bind(wl_resource* new_resource) override;
+
+    std::shared_ptr<IdleInhibitV1Ctx> const ctx;
+};
+
+class IdleInhibitManagerV1
+    : public wayland::IdleInhibitManagerV1
+{
+public:
+    IdleInhibitManagerV1(wl_resource* resource, std::shared_ptr<IdleInhibitV1Ctx> const& ctx);
+
+private:
+    void create_inhibitor(struct wl_resource* id, struct wl_resource* surface) override;
+
+    std::shared_ptr<IdleInhibitV1Ctx> const ctx;
+};
+
+class IdleInhibitorV1
+    : public wayland::IdleInhibitorV1
+{
+public:
+    IdleInhibitorV1(wl_resource* resource, std::shared_ptr<IdleInhibitV1Ctx> const& ctx);
+    ~IdleInhibitorV1();
+
+private:
+    // Not sure why, but removing this unused StateObserver stops the functionality of the IdleInhibit.
+    // It seems unlikely that it will ever actually be used.
+    struct StateObserver : ms::IdleStateObserver
+    {
+        StateObserver(IdleInhibitorV1* idle_inhibitor)
+                : idle_inhibitor{idle_inhibitor}
+        {
+        }
+
+        void idle() override;
+
+        void active() override;
+
+        IdleInhibitorV1* const idle_inhibitor;
+    };
+
+    std::shared_ptr<IdleInhibitV1Ctx> const ctx;
+    std::shared_ptr<StateObserver> const state_observer;
+
+    /// (Would be) called by the state observer
+    /// @{
+    void idle();
+
+    void active();
+    /// @}
+};
+}
+}
+
+auto mf::create_idle_inhibit_manager_v1(
+        wl_display* display,
+        std::shared_ptr<Executor> const& wayland_executor,
+        std::shared_ptr<scene::IdleHub> const& idle_hub)
+-> std::shared_ptr<mw::IdleInhibitManagerV1::Global>
+{
+    auto ctx = std::shared_ptr<IdleInhibitV1Ctx>{new IdleInhibitV1Ctx{wayland_executor, idle_hub}};
```

Comment:
Currently, `IdleInhibitV1Ctx` is a struct with no constructor. This seems to change the behavior of `std::make_shared()` and trying to create a `shared_ptr` with `make_shared` only works if I pass in zero arguments. This follows the pattern as seen in `text_input_v2.cpp`.

If there is a performance/security consideration in this approach, a constructor could be added to `IdleInhibitV1Ctx`.

Reply:
> `std::make_shared<IdleInhibitV1Ctx>(wayland_executor, idle_hub)`

As @graysonguarino says, that doesn't work with lexical types. But adding a trivial constructor would fix it. (If adopting that approach adding a default constructor would avoid surprising semantics.)

> If there is a performance/security consideration in this approach

Consider: constructing the control block in the `shared_ptr<>` could throw. At that point, nothing owns the explicitly `newed` `IdleInhibitV1Ctx` and it will leak (along with anything it owns).

The least typing to get working code:

```suggestion
    auto ctx = std::shared_ptr<IdleInhibitV1Ctx>{IdleInhibitV1Ctx{wayland_executor, idle_hub}};
```

Reply:
This pattern was inherited from my code. I didn't think to pass the struct into `make_shared()` by value. I'll have to update my other protocol impls to do that.